src/webapi/date.rs

```diff
@@ -1,25 +1,903 @@
-use std::marker::PhantomData;
-use webcore::try_from::TryInto;
+use std::collections::HashMap;
+use webcore::try_from::{TryFrom, TryInto};
+use webcore::value::{Value, Reference};
 
-/// [(JavaScript docs)](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Date)
-#[derive(Debug)]
-pub struct Date {
-    dummy: PhantomData< () >
-}
-
-impl Date {
-    /// The Date.now() method returns the number of milliseconds elapsed since 1 January 1970 00:00:00 UTC.
+/// A date object
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)
+pub trait IDate: AsRef< Reference > + TryFrom< Value >  {
+    /// Returns the number of milliseconds elapsed since 1 January 1970 00:00:00 UTC.
     ///
     /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now)
-    pub fn now() -> f64 {
+    #[inline]
+    fn now() -> f64 {
+        js!( return Date.now(); ).try_into().unwrap()
+    }
+
+    /// Accepts the same parameters as the longest form of the constructor,
+    /// and returns the number of milliseconds in a Date object since January 1, 1970, 00:00:00,
+    /// universal time
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/utc)
+    #[inline]
+    fn utc( year: i32, month: u8, day: u8, hour: u8, minutes: u16, seconds: u16, milliseconds: u32) -> i64 {
+        let args_obj: HashMap< &str, Value > = [
+            ("year",         Value::Number( year.into() )),
+            ("month",        Value::Number( month.into() )),
+            ("day",          Value::Number( day.into() )),
+            ("hour",         Value::Number( hour.into() )),
+            ("minutes",      Value::Number( minutes.into() )),
+            ("seconds",      Value::Number( seconds.into() )),
+            ("milliseconds", Value::Number( milliseconds.into() ))
+        ].iter().cloned().collect();
+
         js!(
-            return Date.now();
+            var args = @{args_obj};
+            return Date.UTC( args.year, args.month, args.day, args.hour, args.minutes, args.seconds, args.milliseconds );
         ).try_into().unwrap()
     }
+
+    /// Parses a string representation of a date, and returns the number of
+    /// milliseconds since 1 January 1970 00:00:00 UTC, or NaN if the string is unrecognize or, in
+    /// some cases, contains illegal data values (e.g. 2015-02-31)
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse)
+    #[inline]
+    fn parse( date_string: &str ) -> i64 {
+        js!( return Date.parse( @{date_string} ); ).try_into().unwrap()
+    }
```

Comment:
Looking at the [specs](https://tc39.github.io/ecma262/#sec-date.parse) this function can return `NaN`, which would trigger a panic. We don't want panics, so it'd be better to check for `NaN` and return an `Option`. Also modify the doc comment to reflect this and write a test.